README.md
# gif-encoder [![Build status](https://travis-ci.org/twolfson/gif-encoder.png?branch=master)](https://travis-ci.org/twolfson/gif-encoder)

Streaming [GIF][] encoder

[GIF]: http://en.wikipedia.org/wiki/Graphics_Interchange_Format

This is built as part of the [gifsockets][] project. It is forked from [gif.js][] to allow for a streaming API and performance optimization.

[gifsockets]: https://github.com/twolfson/gifsockets-server

## Getting Started
Install the module with: `npm install gif-encoder`

```js
// Create a 10 x 10 gif
var GifEncoder = require('gif-encoder');
var gif = new GifEncoder(10, 10);

// using an rgba array of pixels [r, g, b, a, ... continues on for every pixel]
// This can be collected from a <canvas> via context.getImageData(0, 0, width, height).data
var pixels = [0, 0, 0, 255/*, ...*/];

// Collect output
var file = require('fs').createWriteStream('img.gif');
gif.pipe(file);

// Write out the image into memory
gif.writeHeader();
gif.addFrame(pixels);
// gif.addFrame(pixels); // Write subsequent rgba arrays for more frames
gif.finish();
```

## Documentation
`gif-encoder` exports `GifEncoder`, a constructor function which extends `readable-stream@~1.1.9`. This means you can use any `streams1`/`streams2` functionality. I will re-iterate what this means below.

```js
// streams1
var gif = new GifEncoder(10, 10);
gif.on('data', console.log);
gif.on('end', process.exit);

// streams2
var gif = new GifEncoder(10, 10);
gif.on('readable', function () {
  console.log(gif.read());
});
```

### `new GifEncoder(width, height, [options])`
Constructor for a new `GifEncoder`

- width `Number` - Width, in pixels, of the `GIF` to output
- height `Number` - Height, in pixels, of the `GIF` to output
- options `Object` - Optional container for any options
    - highWaterMark `Number` - Number, in bytes, to store in internal buffer. Defaults to 64kB.

**NEVER CALL `.removeAllListeners()`. NO DATA EVENTS WILL BE ABLE TO EMIT.**

> We implement the GIF89a specification which can be found at
>
> http://www.w3.org/Graphics/GIF/spec-gif89a.txt

### Events
#### Event: `data`
`function (buffer) {}`

Emits a [`Buffer`][] containing either header bytes, frame bytes, or footer bytes.

[`Buffer`]: http://nodejs.org/api/buffer.html

#### Event: `end`
`function () {}`

Signifies end of the encoding has been reached. This will be emitted once `.finish()` is called.

#### Event: `error`
`function (error) {}`

Emits an `Error` when internal buffer is exceeded. This occurs when you do not `read` (either via `.on('data')` or `.read()`) and we cannot flush prepared data.

> If you have a very large GIF, you can update [`options.highWaterMark`][Constructor] via the [Constructor][].

[Constructor]: #constructor

#### Event: `readable`
`function () {}`

Emits when the stream is ready to be `.read()` from.

#### Event: `writeHeader#start/stop`
`function () {}`

Emits when at the start and end of `.writeHeader()`.

#### Event: `frame#start/stop`
`function () {}`

Emits when at the start and end of `.addFrame()`

#### Event: `finish#start/stop`
`function () {}`

Emits when at the start and end of `.finish()`

### Settings
#### `gif.setDelay(ms)`
Set milliseconds to wait between frames

- ms `Number` - Amount of milliseconds to delay between frames

#### `setFrameRate(framesPerSecond)`
Set delay based on amount of frames per second. Cannot be used with `gif.setDelay`.

- framesPerSecond `Number` - Amount of frames per second

#### `setDispose(disposalCode)`
Set the disposal code

- disposalCode `Number` - Alters behavior of how to render between frames
    - If no transparent color has been set, defaults to 0.
    - Otherwise, defaults to 2.

```
Values :    0 -   No disposal specified. The decoder is
                  not required to take any action.
            1 -   Do not dispose. The graphic is to be left
                  in place.
            2 -   Restore to background color. The area used by the
                  graphic must be restored to the background color.
            3 -   Restore to previous. The decoder is required to
                  restore the area overwritten by the graphic with
                  what was there prior to rendering the graphic.
         4-7 -    To be defined.
```

Taken from http://www.w3.org/Graphics/GIF/spec-gif89a.txt

#### `setRepeat(n)`
Sets amount of times to repeat `GIF`

- n `Number`
    - If `n` is -1, play once.
    - If `n` is 0, loop indefinitely.
    - If `n` is a positive number, loop `n` times.

#### `setTransparent(color)`
Define the color which represents transparency in the `GIF`.

- color `Hexadecimal Number` - Color to represent transparent background
  - Example: `0x00FF00`

#### `setQuality(quality)`
Set the quality (computational/performance trade-off).

- quality `Positive number`
    - 1 is best colors, worst performance.
    - 20 is suggested maximum but there is no limit.
    - 10 is the default, provided an even trade-off.

### Input/output
#### `read([size])`
Read out `size` bytes or until the end of the buffer. This is implemented by `readable-stream`.

- size `Number` - Optional number of bytes to read out

#### `writeHeader()`
Write out header bytes. We are following `GIF89a` specification.

#### `addFrame(imageData, options)`
Write out a new frame to the GIF.

- imageData `Array` - Array of pixels for the new frame. It should follow the sequence of `r, g, b, a` and be `4 * height * width` in length.
    - If used with the options `palette` and `indexedPixels`, then this becomes the index in the palette (e.g. `0` for `color #0`)
- options `Object` - Optional container for options
    - palette `Array` - Array of pixels to use as palette for the frame. It should follow the sequence of `r, g, b, a`
        - At the moment, this must be used with `options.indexedPixels`
    - indexedPixels `Boolean` -  Indicator to treat `imageData` as RGBA values (`false`) or indicies in `palette` (`true`)

#### `finish()`
Write out footer bytes.

### Low-level
For performance in [gifsockets][], we needed to open up some lower level methods for fancy tricks.

**Don't use these unless you know what you are doing.**

#### `flushData()`
We have a secondary internal buffer that collects each byte from `writeByte`. This is to prevent create a new `Buffer` and `data` event for *every byte of data*.

This method empties the internal buffer and pushes it out to the `stream` buffer for reading.

#### `pixels`
Internal store for `imageData` passed in by `addFrame`.

#### `analyzeImage(imageData, options)`
First part of `addFrame`; runs `setImagePixels(removeAlphaChannel(imageData))` and runs `analyzePixels()`.

- imageData `Array` - Same as that in [`addFrame`][]
- options `Object` - Optional container for options
    - indexedPixels `Boolean` -  Indicator to treat `imageData` as RGBA values (`false`) or indicies in `palette` (`true`)

[`addFrame`]: #addframeimagedata

#### `removeAlphaChannel(imageData)`
Reduces `imageData` into a `Uint8Array` of length `3 * width * height` containing sequences of `r, g, b`; removing the alpha channel.

- imageData `Array` - Same as that in [`addFrame`][]; array containing `r, g, b, a` sequences.

#### `setImagePixels(pixels)`
Save `pixels` as `this.pixels` for image analysis.

- pixels `Array` - Same as `imageData` from [`addFrame`][]
    - **`GifEncoder` will mutate the original data.**

#### `setImagePalette(palette)`
Save `palette` as `this.userPalette` for frame writing.

- palette `Array` - Same as `options.palette` from [`addFrame`][]

#### `writeImageInfo()`
Second part of `addFrame`; behavior varies on if it is the first frame or following frame.

In either case, it writes out a bunch of bytes about the image (e.g. palette, color tables).

#### `outputImage()`
Third part of `addFrame`; encodes the analyzed/indexed pixels for the GIF format.

## Donating
Support this project and [others by twolfson][twolfson-projects] via [donations][twolfson-support-me].

<http://twolfson.com/support-me>

[twolfson-projects]: http://twolfson.com/projects
[twolfson-support-me]: http://twolfson.com/support-me

## Contributing
In lieu of a formal styleguide, take care to maintain the existing coding style. Add unit tests for any new or changed functionality. Lint via [grunt](https://github.com/gruntjs/grunt) and test via `npm test`.

## UNLICENSE
As of Nov 11 2013, Todd Wolfson has released all code differences since initial fork from [gif.js][] to the public domain.

These differences have been released under the [UNLICENSE][].

[UNLICENSE]: UNLICENSE

At the [gif.js][] time of forking, [gif.js][] was using the [MIT license][].

[gif.js]: https://github.com/jnordberg/gif.js/tree/faee238491302de05a1ed05e4fbe562738a76310

[MIT license]: https://github.com/jnordberg/gif.js/tree/faee238491302de05a1ed05e4fbe562738a76310#license


CHANGELOG.md
# gif-encoder changelog
0.7.2 - Removed accidental length requirement documentation. Fixes #18

0.7.1 - Fixed supported Node.js versions and Travis CI

0.7.0 - Added support for palette and indexed pixels via @nurpax in #15

0.6.1 - Replaced Gratipay with support me page

0.6.0 - Repaired transparency support via @scinos in #7

0.5.0 - Dropped Node.js@0.8.0 support and updated Travis CI versions

0.4.3 - Removed global variable leak via @mattbierner in #6

0.4.2 - Added foundry for releases

0.4.1 - Introducing performance comparison between byte by byte buffers vs frame based buffers.

0.4.0 - Broke down `getImagePixels` into `removeAlphaChannel` and `setImagePixels`

0.3.0 - Moved to readable-stream for streams1/streams2 functionality

0.2.0 - Moved to larger chunked data events

0.1.0 - Moved to stream-like API (copied from gifsockets-server#f385006)

0.0.3 - Added basic info to README

0.0.2 - Added test for original GIFEncoder functionality

0.0.1 - Initial fork from gif.js#faee238


test/gif-encoder_test.js
var assert = require('assert');
var fs = require('fs');

var async = require('async');
var ndarrayFill = require('ndarray-fill');
var zeros = require('zeros');

var GifEncoder = require('../lib/GIFEncoder.js');
var checkerboardPixels = require('./test-files/checkerboard-pixels.json');

function createGif(height, width, options) {
  before(function () {
    this.gif = new GifEncoder(height, width, options);
  });
}

describe('GifEncoder encoding a checkerboard', function () {
  createGif(10, 10);
  before(function () {
    this.gif.writeHeader();
    this.gif.addFrame(checkerboardPixels);
    this.gif.finish();
  });
  before(function (done) {
    this.gif.once('readable', done);
  });

  it('generates the expected bytes', function () {
    // TODO: Output canvas to a file, perceptual diff GIF to canvas output =3
    // Grab the expected and actual content
    var expectedBytes = fs.readFileSync(__dirname + '/expected-files/checkerboard.gif');
    var actualBytes = this.gif.read();

    // DEV: Write out actual file to expected file
    if (process.env.DEBUG_TEST) {
      try { fs.mkdirSync(__dirname + '/actual-files'); } catch (e) {}
      fs.writeFileSync(__dirname + '/actual-files/checkerboard.gif', actualBytes);
    }

    // Assert the expected matches the actual content
    assert.deepEqual(expectedBytes, actualBytes);
  });
});

describe('GifEncoder encoding a multi-framed checkerboard', function () {
  createGif(10, 10);
  before(function () {
    this.gif.writeHeader();
    this.gif.setDelay(500);
    this.gif.setRepeat(0);
    this.gif.addFrame(checkerboardPixels);
    this.gif.addFrame(require('./test-files/inverse-checkerboard-pixels.json'));
    this.gif.finish();
  });
  before(function (done) {
    this.gif.once('readable', done);
  });

  it('generates the expected bytes', function () {
    var expectedBytes = fs.readFileSync(__dirname + '/expected-files/alternating-checkerboard.gif');
    var actualBytes = this.gif.read();
    if (process.env.DEBUG_TEST) {
      try { fs.mkdirSync(__dirname + '/actual-files'); } catch (e) {}
      fs.writeFileSync(__dirname + '/actual-files/alternating-checkerboard.gif', actualBytes);
    }
    assert.deepEqual(expectedBytes, actualBytes);
  });
});

describe('GifEncoder encoding an overly large, underly read checkerboard', function () {
  createGif(10, 10);
  before(function (done) {
    var that = this;
    this.gif.writeHeader();
    this.gif.on('error', function saveError (err) {
      that.error = err;
    });

    // Write out a new frame until we encounter an error
    // DEV: This is async so mocha can time us out
    async.until(function errorHasOccurred () {
      return that.error;
    }, function addNewFrame(cb) {
      process.nextTick(function () {
        that.gif.addFrame(checkerboardPixels);
        cb();
      });
    }, done);
  });

  it('emits an error', function () {
    assert.notEqual(this.error, undefined);
  });
});

describe('GifEncoder encoding a multi-framed image with a transparent background', function () {
  createGif(15, 15);
  before(function () {
    this.gif.writeHeader();
    this.gif.setDelay(500);
    this.gif.setTransparent(0xFF00FF);
    this.gif.setRepeat(0);
    var i = 0;
    for (; i < 3; i++) {
      // DEV: We are drawing a diagonally moving dot
      // +------+    +------+    +------+
      // |xx    |    |      |    |      |
      // |xx    |    |      |    |      |
      // |      | -> |  xx  | -> |      |
      // |      | -> |  xx  | -> |      |
      // |      |    |      |    |    xx|
      // |      |    |      |    |    xx|
      // +------+    +------+    +------+

      // Generate our frame
      var frameNdarray = zeros([15, 15, 4]);
      ndarrayFill(frameNdarray, function fillFrame (x, y, rgbaIndex) {
        // If this is the alpha channel, always return it as full
        if (rgbaIndex === 3) {
          return 0xFF;
        }

        // Otherwise, if we are on our black dot, then draw it
        if (i * 5 <= x && x < (i + 1) * 5 &&
            i * 5 <= y && y < (i + 1) * 5) {
          // Generate black dot (00 00 00)
          return 0x00;
        // Otherwise, draw our transparent color
        } else {
          return rgbaIndex === 1 ? 0x00 : 0xFF;
        }
      });

      // Draw our frame
      this.gif.addFrame(frameNdarray.data);
    }
    this.gif.finish();
  });
  before(function (done) {
    this.gif.once('readable', done);
  });

  it('generates the expected bytes', function () {
    var expectedBytes = fs.readFileSync(__dirname + '/expected-files/moving-dot.gif');
    var actualBytes = this.gif.read();
    if (process.env.DEBUG_TEST) {
      try { fs.mkdirSync(__dirname + '/actual-files'); } catch (e) {}
      fs.writeFileSync(__dirname + '/actual-files/moving-dot.gif', actualBytes);
    }
    assert.deepEqual(expectedBytes, actualBytes);
  });
});

describe('GifEncoder encoding a checkerboard from indexed pixels', function () {
  createGif(10, 10);
  before(function () {
    this.gif.writeHeader();
    this.gif.addFrame([
      1, 1, 1, 1, 1, 0, 0, 0, 0, 0,
      1, 1, 1, 1, 1, 0, 0, 0, 0, 0,
      1, 1, 1, 1, 1, 0, 0, 0, 0, 0,
      1, 1, 1, 1, 1, 0, 0, 0, 0, 0,
      1, 1, 1, 1, 1, 0, 0, 0, 0, 0,
      0, 0, 0, 0, 0, 1, 1, 1, 1, 1,
      0, 0, 0, 0, 0, 1, 1, 1, 1, 1,
      0, 0, 0, 0, 0, 1, 1, 1, 1, 1,
      0, 0, 0, 0, 0, 1, 1, 1, 1, 1,
      0, 0, 0, 0, 0, 1, 1, 1, 1, 1,
    ], {
      palette: [0x00, 0x00, 0x00, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF],
      indexedPixels: true
    });
    this.gif.finish();
  });
  before(function (done) {
    this.gif.once('readable', done);
  });

  it('generates the expected bytes', function () {
    // TODO: Output canvas to a file, perceptual diff GIF to canvas output =3
    // Grab the expected and actual content
    var expectedBytes = fs.readFileSync(__dirname + '/expected-files/checkerboard-indexed.gif');
    var actualBytes = this.gif.read();

    // DEV: Write out actual file to expected file
    if (process.env.DEBUG_TEST) {
      try { fs.mkdirSync(__dirname + '/actual-files'); } catch (e) {}
      fs.writeFileSync(__dirname + '/actual-files/checkerboard-indexed.gif', actualBytes);
    }

    // Assert the expected matches the actual content
    assert.deepEqual(expectedBytes, actualBytes);
  });
});


test/gif-encoder_benchmark.js
var imageUtils = require('./utils/image');
var GifEncoder = require('../');

describe('A GifEncoder', function () {
  describe('encoding a bunch of frames to `data` events', function () {
    before(function createGifEncoder () {
      this.gif = new GifEncoder(200, 200);
      this.gif.writeHeader();

      // Pipe output to nowhere
      // TODO: Do the same for .read()
      this.gif.on('data', function () {});
    });
    imageUtils.load('medium-size.png');
    before(function encodeABunchOfFrames () {
      var startTime = Date.now();
      var i = 500;
      var gif = this.gif;
      var pixels = this.pixels;
      while (i--) {
        gif.addFrame(pixels);
      }
      var endTime = Date.now();
      this.totalTime = endTime - startTime;
    });

    it('can do so efficiently', function () {
      // TODO: Move to ops/second via benchmarkjs?
      // Medium size x 500 frames 13088 ms for frame based buffers
      // Medium size x 500 frames 14245 ms for byte by byte
      // However, there is a process.nextTick overflow with all the events
      console.log('Medium size x 500 frames', this.totalTime + ' ms');
    });
  });
});

lib/GIFEncoder.js
/*
  GIFEncoder.js

  Authors
  Kevin Weiner (original Java version - kweiner@fmsware.com)
  Thibault Imbert (AS3 version - bytearray.org)
  Johan Nordberg (JS version - code@johan-nordberg.com)
  Todd Wolfson (Implemented streams - todd@twolfson.com)
*/

var assert = require('assert');
var EventEmitter = require('events').EventEmitter;
var ReadableStream = require('readable-stream');
var util = require('util');

var NeuQuant = require('./TypedNeuQuant.js');
var LZWEncoder = require('./LZWEncoder.js');

// DEV: By using a capacitor, we prevent creating a data event for every byte written
function ByteCapacitor(options) {
  // Inherit from ReadableStream
  ReadableStream.call(this, options);

  // Start with an empty buffer and allow writes
  this.okayToPush = true;
  this.resetData();
}
util.inherits(ByteCapacitor, ReadableStream);

ByteCapacitor.prototype._read = function () {
  // The output is controlled by the input provided by methods.
  // If we exceed the highwater mark, we will raise an error.
  this.okayToPush = true;
};

ByteCapacitor.prototype.resetData = function () {
  this.data = [];
};

ByteCapacitor.prototype.flushData = function () {
  // If we are not okay to push, emit an error
  if (!this.okayToPush) {
    var err = new Error('GIF memory limit exceeded. Please `read` from GIF before writing additional frames/information.');
    return this.emit('error', err);
  }

  // Otherwise, push out the new buffer
  var buff = new Buffer(this.data);
  this.resetData();
  this.okayToPush = this.push(buff);
};

ByteCapacitor.prototype.writeByte = function (val) {
  this.data.push(val);
};

ByteCapacitor.prototype.writeUTFBytes = function (string) {
  for (var l = string.length, i = 0; i < l; i++) {
    this.writeByte(string.charCodeAt(i));
  }
};

ByteCapacitor.prototype.writeBytes = function (array, offset, length) {
  for (var l = length || array.length, i = offset || 0; i < l; i++) {
    this.writeByte(array[i]);
  }
};

function GIFEncoder(width, height, options) {
  // Fallback options
  options = options || {};

  // Inherit from ByteCapacitor immediately
  // https://github.com/isaacs/readable-stream/blob/v1.1.9/lib/_stream_readable.js#L60-L63
  var hwm = options.highWaterMark;
  ByteCapacitor.call(this, {
    // Allow for up to 64kB of GIFfy-goodness
    highWaterMark: (hwm || hwm === 0) ? hwm : 64 * 1024
  });

  // image size
  this.width = ~~width;
  this.height = ~~height;

  // transparent color if given
  this.transparent = null;

  // transparent index in color table
  this.transIndex = 0;

  // -1 = no repeat, 0 = forever. anything else is repeat count
  this.repeat = -1;

  // frame delay (hundredths)
  this.delay = 0;

  this.pixels = null; // BGR byte array from frame
  this.indexedPixels = null; // converted frame indexed to palette
  this.colorDepth = null; // number of bit planes
  this.colorTab = null; // RGB palette
  this.userPalette = null; // User-input palette
  this.usedEntry = []; // active palette entries
  this.palSize = 7; // color table size (bits-1)
  this.dispose = -1; // disposal code (-1 = use default)
  this.firstFrame = true;
  this.sample = 10; // default sample interval for quantizer

  // When we encounter a header, new frame, or stop, emit data
  var that = this;
  function flushData() {
    that.flushData();
  }
  this.on('writeHeader#stop', flushData);
  this.on('frame#stop', flushData);
  this.on('finish#stop', function finishGif () {
    // Flush the data
    flushData();

    // Close the gif
    that.push(null);
  });
}
util.inherits(GIFEncoder, ByteCapacitor);

/*
  Sets the delay time between each frame, or changes it for subsequent frames
  (applies to last frame added)
*/
GIFEncoder.prototype.setDelay = function(milliseconds) {
  this.delay = Math.round(milliseconds / 10);
};

/*
  Sets frame rate in frames per second.
*/
GIFEncoder.prototype.setFrameRate = function(fps) {
  this.delay = Math.round(100 / fps);
};

/*
  Sets the GIF frame disposal code for the last added frame and any
  subsequent frames.

  Default is 0 if no transparent color has been set, otherwise 2.
*/
GIFEncoder.prototype.setDispose = function(disposalCode) {
  if (disposalCode >= 0) this.dispose = disposalCode;
};

/*
  Sets the number of times the set of GIF frames should be played.

  -1 = play once
  0 = repeat indefinitely

  Default is -1

  Must be invoked before the first image is added
*/

GIFEncoder.prototype.setRepeat = function(repeat) {
  this.repeat = repeat;
};

/*
  Sets the transparent color for the last added frame and any subsequent
  frames. Since all colors are subject to modification in the quantization
  process, the color in the final palette for each frame closest to the given
  color becomes the transparent color for that frame. May be set to null to
  indicate no transparent color.
*/
GIFEncoder.prototype.setTransparent = function(color) {
  this.transparent = color;
};

// Custom methods for performance hacks around streaming GIF data pieces without re-analyzing/loading
GIFEncoder.prototype.analyzeImage = function (imageData, options) {
  // convert to correct format if necessary
  if (options && options.palette) {
    this.setImagePalette(this.removeAlphaChannel(options.palette));
  }
  if (options && options.indexedPixels === true) {
    assert(options.palette, '`options.indexedPixels` requires `options.palette` to load from. Please include one`');
    this.setImagePixels(imageData);
  } else {
    this.setImagePixels(this.removeAlphaChannel(imageData));
  }
  this.analyzePixels(options); // build color table & map pixels
  if (options && options.palette) {
    this.userPalette = null;
  }
};

GIFEncoder.prototype.writeImageInfo = function () {
  if (this.firstFrame) {
    this.writeLSD(); // logical screen descriptior
    this.writePalette(); // global color table
    if (this.repeat >= 0) {
      // use NS app extension to indicate reps
      this.writeNetscapeExt();
    }
  }

  this.writeGraphicCtrlExt(); // write graphic control extension
  this.writeImageDesc(); // image descriptor
  if (!this.firstFrame) this.writePalette(); // local color table

  // DEV: This was originally after outputImage but it does not affect order it seems
  this.firstFrame = false;
};

GIFEncoder.prototype.outputImage = function () {
  this.writePixels(); // encode and write pixel data
};

/*
  Adds next GIF frame. The frame is not written immediately, but is
  actually deferred until the next frame is received so that timing
  data can be inserted.  Invoking finish() flushes all frames.
*/
GIFEncoder.prototype.addFrame = function(imageData, options) {
  this.emit('frame#start');

  this.analyzeImage(imageData, options);
  this.writeImageInfo();
  this.outputImage();

  this.emit('frame#stop');
};

/*
  Adds final trailer to the GIF stream, if you don't call the finish method
  the GIF stream will not be valid.
*/
GIFEncoder.prototype.finish = function() {
  this.emit('finish#start');
  this.writeByte(0x3b); // gif trailer
  this.emit('finish#stop');
};

/*
  Sets quality of color quantization (conversion of images to the maximum 256
  colors allowed by the GIF specification). Lower values (minimum = 1)
  produce better colors, but slow processing significantly. 10 is the
  default, and produces good color mapping at reasonable speeds. Values
  greater than 20 do not yield significant improvements in speed.
*/
GIFEncoder.prototype.setQuality = function(quality) {
  if (quality < 1) quality = 1;
  this.sample = quality;
};

/*
  Writes GIF file header
*/
GIFEncoder.prototype.writeHeader = function() {
  this.emit('writeHeader#start');
  this.writeUTFBytes("GIF89a");
  this.emit('writeHeader#stop');
};

/*
  Analyzes current frame colors and creates color map.
*/
GIFEncoder.prototype.analyzePixels = function(options) {
  // If we're being called with a user defined palette, skip NeuQuant and
  // color remapping.
  var nPix, index, imgq;
  if (this.userPalette !== null) {
    assert(options && options.indexedPixels === true,
      '`palette` can only be used with `options.indexedPixels` at the moment. ' +
      'Please add `{indexedPixels: true}` to `addFrame()`');
    nPix = this.pixels.length;
    this.indexedPixels = new Uint8Array(nPix);

    this.colorTab = this.userPalette;

    for (var i = 0; i < nPix; i++) {
      index = this.pixels[i];
      this.indexedPixels[i] = index;
      this.usedEntry[index] = true;
    }
  } else {
    var len = this.pixels.length;
    nPix = len / 3;
    this.indexedPixels = new Uint8Array(nPix);

    imgq = new NeuQuant(this.pixels, this.sample);
    imgq.buildColormap(); // create reduced palette
    this.colorTab = imgq.getColormap();

    // map image pixels to new palette
    var k = 0;
    for (var j = 0; j < nPix; j++) {
      index = imgq.lookupRGB(
        this.pixels[k++] & 0xff,
        this.pixels[k++] & 0xff,
        this.pixels[k++] & 0xff
      );
      this.usedEntry[index] = true;
      this.indexedPixels[j] = index;
    }
  }

  this.pixels = null;
  this.colorDepth = 8;
  this.palSize = 7;

  // find index for transparent color
  if (this.transparent !== null) {
    this.transIndex = imgq.lookupRGB(
      (this.transparent & 0xFF0000) >> 16,
      (this.transparent & 0x00FF00) >> 8,
      (this.transparent & 0x0000FF)
    );
  } else {
    this.transIndex = 0;
  }
};


/*
  Extracts image pixels into byte array pixels
  (removes alphachannel from canvas imagedata)
*/
GIFEncoder.prototype.removeAlphaChannel = function (data) {
  var w = this.width;
  var h = this.height;
  var pixels = new Uint8Array(w * h * 3);

  var count = 0;

  for (var i = 0; i < h; i++) {
    for (var j = 0; j < w; j++) {
      var b = (i * w * 4) + j * 4;
      pixels[count++] = data[b];
      pixels[count++] = data[b+1];
      pixels[count++] = data[b+2];
    }
  }

  return pixels;
};

GIFEncoder.prototype.setImagePixels = function(pixels) {
  this.pixels = pixels;
};
GIFEncoder.prototype.setImagePalette = function(userPalette) {
  this.userPalette = userPalette;
};

/*
  Writes Graphic Control Extension
*/
GIFEncoder.prototype.writeGraphicCtrlExt = function() {
  this.writeByte(0x21); // extension introducer
  this.writeByte(0xf9); // GCE label
  this.writeByte(4); // data block size

  var transp, disp;
  if (this.transparent === null) {
    transp = 0;
    disp = 0; // dispose = no action
  } else {
    transp = 1;
    disp = 2; // force clear if using transparent color
  }

  if (this.dispose >= 0) {
    disp = this.dispose & 7; // user override
  }
  disp <<= 2;

  // packed fields
  this.writeByte(
    0 | // 1:3 reserved
    disp | // 4:6 disposal
    0 | // 7 user input - 0 = none
    transp // 8 transparency flag
  );

  this.writeShort(this.delay); // delay x 1/100 sec
  this.writeByte(this.transIndex); // transparent color index
  this.writeByte(0); // block terminator
};

/*
  Writes Image Descriptor
*/
GIFEncoder.prototype.writeImageDesc = function() {
  this.writeByte(0x2c); // image separator
  this.writeShort(0); // image position x,y = 0,0
  this.writeShort(0);
  this.writeShort(this.width); // image size
  this.writeShort(this.height);

  // packed fields
  if (this.firstFrame) {
    // no LCT - GCT is used for first (or only) frame
    this.writeByte(0);
  } else {
    // specify normal LCT
    this.writeByte(
      0x80 | // 1 local color table 1=yes
      0 | // 2 interlace - 0=no
      0 | // 3 sorted - 0=no
      0 | // 4-5 reserved
      this.palSize // 6-8 size of color table
    );
  }
};

/*
  Writes Logical Screen Descriptor
*/
GIFEncoder.prototype.writeLSD = function() {
  // logical screen size
  this.writeShort(this.width);
  this.writeShort(this.height);

  // packed fields
  this.writeByte(
    0x80 | // 1 : global color table flag = 1 (gct used)
    0x70 | // 2-4 : color resolution = 7
    0x00 | // 5 : gct sort flag = 0
    this.palSize // 6-8 : gct size
  );

  this.writeByte(0); // background color index
  this.writeByte(0); // pixel aspect ratio - assume 1:1
};

/*
  Writes Netscape application extension to define repeat count.
*/
GIFEncoder.prototype.writeNetscapeExt = function() {
  this.writeByte(0x21); // extension introducer
  this.writeByte(0xff); // app extension label
  this.writeByte(11); // block size
  this.writeUTFBytes('NETSCAPE2.0'); // app id + auth code
  this.writeByte(3); // sub-block size
  this.writeByte(1); // loop sub-block id
  this.writeShort(this.repeat); // loop count (extra iterations, 0=repeat forever)
  this.writeByte(0); // block terminator
};

/*
  Writes color table
*/
GIFEncoder.prototype.writePalette = function() {
  this.writeBytes(this.colorTab);
  var n = (3 * 256) - this.colorTab.length;
  for (var i = 0; i < n; i++)
    this.writeByte(0);
};

GIFEncoder.prototype.writeShort = function(pValue) {
  this.writeByte(pValue & 0xFF);
  this.writeByte((pValue >> 8) & 0xFF);
};

/*
  Encodes and writes pixel data
*/
GIFEncoder.prototype.writePixels = function() {
  var enc = new LZWEncoder(this.width, this.height, this.indexedPixels, this.colorDepth);
  enc.encode(this);
};

/*
  Retrieves the GIF stream
*/
GIFEncoder.prototype.stream = function() {
  return this;
};

GIFEncoder.ByteCapacitor = ByteCapacitor;

module.exports = GIFEncoder;
